File: bin_search.go
package main

import (
	"bytes"
	"fmt"
)

type Nameval struct {
	name        string
	value       int
	left, right *Nameval
}

func (treep *Nameval) lookup(name string) *Nameval {
	if treep == nil {
		return nil
	}
	cmp := bytes.Compare([]byte(name), []byte(treep.name))
	if cmp == 0 {
		return treep
	} else if cmp < 0 {
		return treep.left.lookup(name)
	} else {
		return treep.right.lookup(name)
	}
}

func (treep *Nameval) insert(newp *Nameval) *Nameval {
	if treep == nil {
		return newp
	}
	cmp := bytes.Compare([]byte(newp.name), []byte(treep.name))
	if cmp == 0 {
		return treep
	} else if cmp < 0 {
		treep.left = treep.left.insert(newp)
	} else {
		treep.right = treep.right.insert(newp)
	}
	return treep
}

func (treep *Nameval) Print() {
	if treep == nil {
		return
	}
	treep.left.Print()
	fmt.Printf("My name is %s\n", treep.name)
	treep.right.Print()
}

func main() {
	foo := &Nameval{name: "foo"}
	bar := &Nameval{name: "bar"}
	derp := &Nameval{name: "derp"}
	eleph := &Nameval{name: "eleph"}
	foo.insert(bar)
	foo.insert(derp)
	foo.insert(eleph)
	foo.Print()
	foo.lookup("derp").Print()
}

File: bin_search.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


typedef struct Nameval Nameval;

struct Nameval {
  char *name;
  int value;
  Nameval *left, *right;
};

Nameval *lookup(Nameval *treep, char *name) {
  if (treep == NULL) {
    return NULL;
  }
  int cmp = strcmp(name, treep->name);
  if (cmp == 0) {
    return treep;
  } else if (cmp < 0) {
    return lookup(treep->left, name);
  } else {
    return lookup(treep->right, name);
  }
}

Nameval *insert(Nameval *treep, Nameval *newp) {
  if (treep == NULL) {
    treep = newp;
    return treep;
  }
  int cmp = strcmp(newp->name, treep->name);
  if (cmp == 0) {
    return treep;
  } else if (cmp < 0) {
    treep->left = insert(treep->left, newp);
  } else {
    treep->right = insert(treep->right, newp);
  }
  return treep;
}

void printTree(Nameval* treep) {
  if (treep == NULL) {
    return;
  }
  printTree(treep->left);
  printf("My name is %s\n", treep->name);
  printTree(treep->right);
}

int main() {
  Nameval foo = (Nameval) {.name = "foo"};
  Nameval bar = (Nameval) {.name = "bar"};
  Nameval derp = (Nameval) {.name = "derp"};
  Nameval eleph = (Nameval) {.name = "elep"};
  insert(&foo, &bar);
  insert(&foo, &derp);
  insert(&foo, &eleph);
  printTree(&foo);
  printTree(lookup(&foo, "derp"));
}
